### DISCUSSION TITLE
`setValue` doesn't update an array properly

### DISCUSSION THREAD

**Discussion by JuanM04** [react-hook-form/react-hook-form#5769] [Q&A] [answered]
**Describe the bug**
Updating arrays with `setValue` works in a weird fashion when there are a default array.

**To Reproduce**

1. Make a form.
2. Add a array value like `number[]`.
3. Add a default value to it, like `[3,1,4]`.
4. Use `setValue` and `watch` to update the values (check out the Codesanbox for a better example). 

**Codesandbox link**
https://codesandbox.io/s/elastic-chebyshev-2jlxu?file=/src/Forms.tsx

**Desktop (please complete the following information):**

- OS: Fedora 34
- Browser: Brave (Chromium-based)
- Version: 91

**Additional context**
Previously mentioned [here](https://github.com/react-hook-form/react-hook-form/issues/5724#issuecomment-867706511).

**Comment by JamesWong1999** [comment]
What's the steps to reproduce the issue? and what's the expected result?

**Comment by bluebill1049** [comment]
> What's the steps to reproduce the issue? and what's the expected result?

@JuanM04 can you elaborate on the above?

**Comment by bluebill1049** [comment, marked as answer]
## What happend?

here is the difference with and without `defaultValues` (without an actual input `register`)

<img width="917" alt="Screen Shot 2021-06-29 at 5 12 37 pm" src="https://user-images.githubusercontent.com/10513364/123753457-35733400-d8fd-11eb-9b12-2489c8c2bc2c.png">

as you can see above, if you didn't register an input but leave it in the `defaultValues`, then react hook form will register those inputs on your behalf which may not be what you desired. eg:

`test: [1,2,3]` could be a single input or three of them.  in your case, it's actually a single input, but the hook form will register 3 of them. 

## How to solve this?

If you have a group input, it's always better to register the input and let hook form know it's a group input instead of multiple inputs.

```tsx
// test: [1,2,3] as defaultValue

useEffect(() => {
  register('test')
}, [register]);
```

hope the above answers your question/issue.

**Reply by JuanM04** [reply]
Thank you for that detailed answer! That solves it perfectly

**Reply by SergeyWebDeveloper** [reply]
thanks bro, faced the same problem!

**Reply by tatane616** [reply]
You saved my day!
